extension/WebJobs.Extensions.RabbitMQ/Trigger/RabbitMQListener.cs

```diff
@@ -20,126 +19,134 @@ namespace Microsoft.Azure.WebJobs.Extensions.RabbitMQ;
+        this.Descriptor = new ScaleMonitorDescriptor($"{functionId}-RabbitMQTrigger-{queueName}");
+        this.logDetails = $"function: '{functionId}', queue: '{queueName}'";
     }
 
     public ScaleMonitorDescriptor Descriptor { get; }
 
     public void Cancel()
     {
-        if (!this.started)
-        {
-            return;
-        }
-
         this.StopAsync(CancellationToken.None).Wait();
     }
 
     public void Dispose()
     {
+        // Nothing to dispose.
     }
 
     public Task StartAsync(CancellationToken cancellationToken)
     {
-        this.ThrowIfDisposed();
+        int previousState = Interlocked.CompareExchange(ref this.listenerState, ListenerStarting, ListenerNotStarted);
 
-        if (this.started)
+        // It is possible that the WebJobs SDKS invokes StartAsync() method more than once, if there are other trigger
+        // listeners registered and some of them have failed to start.
+        if (previousState != ListenerNotStarted)
         {
-            throw new InvalidOperationException("The listener has already been started.");
+            throw new InvalidOperationException("The listener is either starting or has already started.");
         }
 
-        this.rabbitMQModel.BasicQos(0, this.prefetchCount, false);  // Non zero prefetchSize doesn't work (tested upto 5.2.0) and will throw NOT_IMPLEMENTED exception
-        this.consumer = new EventingBasicConsumer(this.rabbitMQModel.Model);
+        // The RabbitMQ server (v3.11.2 as of latest) only has support for prefetch size of zero (no specific limit).
+        // See: https://github.com/rabbitmq/rabbitmq-server/blob/v3.11.2/deps/rabbit/src/rabbit_channel.erl#L1543.
+        // See: https://www.rabbitmq.com/amqp-0-9-1-reference.html#basic.qos.prefetch-size for protocol specification.
+        this.channel.BasicQos(prefetchSize: 0, this.prefetchCount, global: false);
 
-        this.consumer.Received += async (model, args) =>
-        {
-            // The RabbitMQ client rents an array from the ArrayPool to hold a copy of the message body, and passes it
-            // to the listener. Once all event handlers are executed, the array is returned back to the pool so that the
-            // memory can be reused for future messages for that connection. However, since our event handler is async,
-            // the very first await statement i.e. the call to TryExecuteAsync below causes the event handler invocation
-            // to complete and lets the RabbitMQ client release the memory. This led to message body corruption when the
-            // message is republished (see: https://github.com/Azure/azure-functions-rabbitmq-extension/issues/211).
-            //
-            // We chose to copy the message body instead of having a new 'args' object as there is only one event
-            // handler registered for the consumer so there should be no side-effects.
-            args.Body = args.Body.ToArray();
```

Comment:
I found that copying message body is not required if we replace `EventingBasicConsumer` with `AsyncEventingBasicConsumer`. Please check my test code: https://gist.github.com/JatinSanghvi/0993825524957bb3f36f207ac7d4a676. The code copies the implementation of `AsyncEventingBasicConsumer` from RabbitMQ client.